File: src/#1008/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func bstFromPreorder(preorder []int) *TreeNode {
	root := &TreeNode{
		Val:   preorder[0],
		Left:  nil,
		Right: nil,
	}
	preorder = preorder[1:]
	for _, v := range preorder {
		newNode := &TreeNode{
			Val:   v,
			Left:  nil,
			Right: nil,
		}
		insert(root, newNode)
	}
	return root
}

func insert(root, node *TreeNode) {
	if node.Val > root.Val {
		if root.Right == nil {
			root.Right = node
			return
		}
		insert(root.Right, node)
	} else {
		if root.Left == nil {
			root.Left = node
			return
		}
		insert(root.Left, node)
	}
}

File: src/#1008/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
}

var bstFromPreorder = function (preorder) {
  function traversal(arr) {
    if (arr.length < 1) return null

    const root = new TreeNode(arr[0])

    if (arr.length == 1) return root
    let firstMax = arr.findIndex(n => n > root.val)
    if (firstMax == -1) firstMax = arr.length


    root.left = traversal(arr.slice(1, firstMax))
    root.right = traversal(arr.slice(firstMax))
    
    return root
  }

  return traversal(preorder)
}
